codegen/src/api/metadata.rs

```diff
@@ -0,0 +1,229 @@
+    sp_core::hashing::sha2_256(bytes)
+}
+
+fn get_field_hash(
+    registry: &PortableRegistry,
+    field: &Field<PortableForm>,
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::Field as u8];
+
+    bytes.extend(field.name().encode());
+    bytes.extend(field.type_name().encode());
+    bytes.extend(get_type_hash(registry, field.ty().id(), set));
+
+    hash(&bytes)
+}
+
+fn get_variant_hash(
+    registry: &PortableRegistry,
+    var: &Variant<PortableForm>,
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::Variant as u8];
+
+    bytes.extend(var.name().as_bytes());
+    for field in var.fields() {
+        bytes.extend(get_field_hash(registry, field, set));
+    }
+
+    hash(&bytes)
+}
+
+fn get_type_def_hash(
+    registry: &PortableRegistry,
+    ty_def: &TypeDef<PortableForm>,
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::TypeDef as u8];
+
+    let data = match ty_def {
+        TypeDef::Composite(composite) => {
+            let mut bytes = Vec::new();
+            for field in composite.fields() {
+                bytes.extend(get_field_hash(registry, field, set));
+            }
+            bytes
+        }
+        TypeDef::Variant(variant) => {
+            let mut bytes = Vec::new();
+            for var in variant.variants() {
+                bytes.extend(get_variant_hash(registry, var, set));
+            }
+            bytes
+        }
+        TypeDef::Sequence(sequence) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(registry, sequence.type_param().id(), set));
+            bytes
+        }
+        TypeDef::Array(array) => {
+            let mut bytes = Vec::new();
+            bytes.extend(array.len().to_be_bytes());
+            bytes.extend(get_type_hash(registry, array.type_param().id(), set));
+            bytes
+        }
+        TypeDef::Tuple(tuple) => {
+            let mut bytes = Vec::new();
+            for field in tuple.fields() {
+                bytes.extend(get_type_hash(registry, field.id(), set));
+            }
+            bytes
+        }
+        TypeDef::Primitive(primitive) => {
+            let mut bytes = Vec::new();
+            bytes.extend(primitive.encode());
+            bytes
+        }
+        TypeDef::Compact(compact) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(registry, compact.type_param().id(), set));
+            bytes
+        }
+        TypeDef::BitSequence(bitseq) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(registry, bitseq.bit_order_type().id(), set));
+            bytes.extend(get_type_hash(registry, bitseq.bit_store_type().id(), set));
+            bytes
+        }
+    };
+    bytes.extend(data);
+    hash(&bytes)
+}
+
+fn get_type_hash(
+    registry: &PortableRegistry,
+    id: u32,
+    set: &mut HashSet<u32>,
```

Comment:
Nit: how about renaming `set` to something more descriptive like, hrm, `visited_ids`?

Reply:
Yep, indeed this needs a refactoring :D Thanks